File: dtrules/src/main/java/com/dtrules/mapping/Mapping.java
/** 
 * Copyright 2004-2011 DTRules.com, Inc.
 * 
 * See http://DTRules.com for updates and documentation for the DTRules Rules Engine  
 *   
 * Licensed under the Apache License, Version 2.0 (the "License");  
 * you may not use this file except in compliance with the License.  
 * You may obtain a copy of the License at  
 *   
 *      http://www.apache.org/licenses/LICENSE-2.0  
 *   
 * Unless required by applicable law or agreed to in writing, software  
 * distributed under the License is distributed on an "AS IS" BASIS,  
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
 * See the License for the specific language governing permissions and  
 * limitations under the License.  
 **/
  
package com.dtrules.mapping;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.HashMap;

import com.dtrules.entity.IREntity;
import com.dtrules.infrastructure.RulesException;
import com.dtrules.interpreter.RName;
import com.dtrules.session.DTState;
import com.dtrules.session.IStreamSource.FileType;
import com.dtrules.session.IRSession;
import com.dtrules.session.RSession;
import com.dtrules.session.RuleSet;
import com.dtrules.session.RulesDirectory;
import com.dtrules.xmlparser.GenericXMLParser;

/**
 * The Mapping class is used to load XML formated data into an instance of
 * an EntityFactory.  
 * @author Paul Snow
 *
 */
public class Mapping {
	
    private final DTState state;
	 	 
    private final IRSession session;
     	
    HashMap<RName,EntityInfo>               entities      = new HashMap<RName,EntityInfo>();
    HashMap<String,String>                  multiple      = new HashMap<String,String>();   // Specifies a tag that uses an attribute to specify what entity to create.
	HashMap<String,EntityInfo>              requests      = new HashMap<String,EntityInfo>(); 
	HashMap<String,String>                  entityinfo	  = new HashMap<String,String>();
	HashMap<String,ArrayList<DataObjectMap>>dataObjects   = new HashMap<String,ArrayList<DataObjectMap>>();
    /**
     * List of attribute/list pairs.  Every entity that defines the given attribute has every instance saved in the given list.
     * Each entry is an array of two strings, the attribute and the list name.
     */
    ArrayList<String[]> attribute2listPairs = new ArrayList<String[]>();
    
	/**
	 * A List of entities that should be created at initialization and put on the Entity Stack.
	 */
	ArrayList<String> entitystack   = new ArrayList<String>(); // Specifies what Entities should be on the Entity Stack to begin with.
   
	
	/**
	 * This is the set of Attribute objects used to map XML tags to Attributes on Entities.
	 */ 
	HashMap<String,AttributeInfo>   setattributes = new HashMap<String,AttributeInfo>();   // Sets an attribute.

    
    private Mapping(IRSession session){
        this.session = session;
        this.state   = session.getState();
    }
    
    private void initialize(){
        for(String entity : entitystack){
            try {
                IREntity newEntity = ((RSession)session).createEntity(null,entity);
                session.getState().entitypush(newEntity);
            } catch (RulesException e) {
                throw new RuntimeException("Failed to initialize the Entity Stack: "+e);
            }
        }
    }
    
    /**
     * For internal use only.  If you need a Mapping object, get it from the appropriate
     * RuleSet object.
     * @param rd
     * @param session
     * @param filename
     * @return
     */
    public static Mapping newMapping(RulesDirectory rd, IRSession session, String filename){
        Mapping mapping = new Mapping(session);
    	try {
			InputStream s = session.getRulesDirectory().getFileSource().openStreamSearch(
			        FileType.MAP, session.getRuleSet(), filename);
            session.getState().traceTagBegin("loadMapping", "file",filename);
			mapping.loadMap(s);
			session.getState().traceTagEnd(); 
		} catch (FileNotFoundException e) {
 			throw new RuntimeException(e);
		} catch (Exception e) {
 			throw new RuntimeException("Error accessing '"+filename+"'\r\n"+e);
		}
		return mapping;
    }
    
    
    /**
     * Clones the mapping structures for use by a given session
     * @param session
     * @param amapping
     * @return
     */
    public Mapping clone(IRSession session){
    	Mapping mapping = new Mapping(session);
    	mapping.entities = entities;
    	mapping.multiple = multiple;
    	mapping.requests = requests;
    	mapping.entityinfo = entityinfo;
    	mapping.entitystack = entitystack;
    	mapping.setattributes = setattributes;
    	mapping.attribute2listPairs = attribute2listPairs;
    	mapping.dataObjects = dataObjects;
    	mapping.initialize();
    	return mapping;
    }
    
    
    /**
     * Constructor for creating a Mapping.  This has been depreciated because
     * building a map this way reloads the mapping file over and over.  Instead
     * you should use:
     * 
     *    Mapping map = session.getMapping();
     * 
     * @param rd
     * @param session
     * @param rs
     * @deprecated
     */
     public Mapping(RulesDirectory rd,IRSession session, RuleSet rs){
       this.session = session;
       this.state   = session.getState();
       String filename = rs.getMapPath().get(0);
       try {
           InputStream s = session.getRulesDirectory().getFileSource().openStreamSearch(
                   FileType.MAP, session.getRuleSet(), filename);
           session.getState().traceTagBegin("loadMapping", "file",filename);
           this.loadMap(s);
           session.getState().traceTagEnd(); 
       } catch (FileNotFoundException e) {
           throw new RuntimeException(e);
       } catch (Exception e) {
           throw new RuntimeException(e);
       }
    }
    
    /**
	 * Load parses XML to EDD Map file into a set of actions
	 * which can be used to build an EDD file.
	 * <br><br>
	 * 
	 * @param file
	 */
	 protected void loadMap(InputStream xmlStream) throws Exception {
		LoadMap map = new LoadMap(state,this);
		GenericXMLParser.load(xmlStream,map);
		if(!map.isLoadSuccessful()){
		    
			throw new Exception("Map failed to load due to errors!");
		}
	} 
	/**
	 * Load Data into the XML from a String.  
	 * 
	 * @param session - The session that the data needs to be loaded into
	 * @param Source - This is for error reporting.  What is the source of the data?
	 * @param data - The XML data as a String.
	 * 
	 * @throws RulesException
	 */
	public void loadStringData(IRSession session, String Source, String data)throws RulesException {
	    Reader strReader = new StringReader(data);
	    loadData(session,strReader, Source);
	}
	 
	public void loadData(IRSession session, String dataSource )throws RulesException {
	    loadData(session, dataSource, dataSource);
	}	      
	
	public void loadData(IRSession session, String dataSource, String source )throws RulesException {
        InputStream input = session.getRulesDirectory().getFileSource().openStreamSearch(
                FileType.MAP, session.getRuleSet(), dataSource);
        if(input == null){
            throw new RulesException("File Not Found","Mapping.loadData()","Could not open "+dataSource);
        }
        loadData(session,input,source);
    }
    
    /**
     * This is the state for dataloading... So that data can be split over multiple files, we
     * cache our LoadXMLData object between calls to loadData().
     */
    LoadXMLData dataloader = null;
    
    /**
     * Loads data using this mapping from an XML data Source into the given session.
     * @param session
     * @param dataSource
     * @throws RulesException
     */
    public void loadData (IRSession session, InputStream dataSource, String source) throws RulesException {
       Reader dataSrc = new InputStreamReader(dataSource);
       loadData(session, dataSrc, source);
    }
    public void loadData (IRSession session, Reader dataSource, String source) throws RulesException {
            if(dataloader ==null) {
            dataloader = new LoadXMLData(this,session,session.getRuleSet().getName());
        }
        if(source==null){
            source = "XML file";
        }
        try {
            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagBegin("dataLoad","source",source);   
            }
            
            GenericXMLParser.load(dataSource, dataloader);

            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagEnd();
            }
        } catch (Exception e) {
            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagEnd();
            }
            throw new RulesException("Parse Error","LoadMap.loadData()",e.getMessage());
        }
    }
    
    /**
     * Load an in Memory data structure into the given session.
     * @param session
     * @param datasrc
     * @throws RulesException
     */
    public void loadData(IRSession session, DataMap datasrc)throws RulesException {
        if(dataloader ==null) {
            dataloader = new LoadDatamapData(this,session,session.getRuleSet().getName());
        }
        try {
            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagBegin("loadData");
            }
            XMLNode tag = datasrc.getRootTag();
            processTag(tag);
            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagEnd();
            }
        }catch(Exception e){
            if(session.getState().testState(DTState.TRACE)){
                session.getState().traceTagEnd();
            }
            throw new RulesException("Parse Error","LoadMap.loadData()",e.getMessage());
        }
        
    }
    String tagstk[]  = new String[1000];
    int    tagstkptr = 0;
    /**
     * Load a tag into a session.
     */
    void processTag(XMLNode tag) throws Exception {
        if(tag.type()== XMLNode.Type.TAG){
            state.traceTagBegin(tag.getTag()==null?"process":tag.getTag(),null);
            tagstk[tagstkptr++] = tag.getTag();
            
            dataloader.beginTag(tagstk, tagstkptr, tag.getTag(), tag.getAttribs());
            if(tag.getTags()!= null) for(XMLNode nextTag : tag.getTags()){
                processTag(nextTag);
            }
            ((LoadDatamapData)dataloader).endTag(tagstk, tagstkptr, tag, tag.getBody(), tag.getAttribs());
            
            tag.clearRef();
 
            tagstkptr--;
            tagstk[tagstkptr]= null;
            state.traceTagEnd();
        }
    }


    /**
     * @return the state
     */
    public DTState getState() {
        return state;
    }


    /**
     * @return the session
     */
    public IRSession getSession() {
        return session;
    }
    
    
}

File: go/pkg/dtrules/mapping/mapping.go
// Copyright 2004-2011 DTRules.com, Inc.
// Copyright 2024 Paul Snow
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package mapping implements the DTRules XML to Entity mapping system.
// It loads mapping definitions and uses them to parse input data XML
// into DTRules entities.
package mapping

import (
	"fmt"
	"io"
	"strings"

	"github.com/PaulSnow/DTRules/go/pkg/dtrules"
)

// Mapping holds the configuration for mapping XML data to entities.
type Mapping struct {
	session dtrules.Session
	state   dtrules.State

	// entities maps entity name to cardinality ("1", "*", or "+")
	entities map[string]string

	// requests maps XML tag to EntityInfo for entity creation
	requests map[string]*EntityInfo

	// setattributes maps XML tag to AttributeInfo for attribute setting
	setattributes map[string]*AttributeInfo

	// multiple maps XML tag to attribute name for polymorphic entity creation
	multiple map[string]string

	// entitystack is the list of entities to create at initialization
	entitystack []string

	// attribute2listPairs maps attributes to lists (for addalltolist)
	attribute2listPairs [][2]string
}

// EntityInfo holds information about an entity to create.
type EntityInfo struct {
	Name   string // entity name
	ID     string // attribute to use as ID
	List   string // list to add entity to
	Entity string // original entity name from mapping
}

// AttributeInfo holds information about setting an attribute.
type AttributeInfo struct {
	Tag     string
	Attribs map[string]*Attrib // keyed by enclosure entity name (empty string for default)
}

// Attrib represents a single attribute mapping.
type Attrib struct {
	Tag        string
	Entity     string
	RAttribute string
	Type       AttributeType
}

// AttributeType represents the type of an attribute.
type AttributeType int

const (
	TypeNone AttributeType = iota
	TypeString
	TypeInteger
	TypeDouble
	TypeBoolean
	TypeDate
	TypeEntity
	TypeArray
	TypeXMLValue
)

// ParseAttributeType converts a string type to AttributeType.
func ParseAttributeType(s string) AttributeType {
	switch strings.ToLower(s) {
	case "string":
		return TypeString
	case "integer", "int":
		return TypeInteger
	case "double", "float", "real":
		return TypeDouble
	case "boolean", "bool":
		return TypeBoolean
	case "date":
		return TypeDate
	case "entity":
		return TypeEntity
	case "array", "list":
		return TypeArray
	case "xmlvalue":
		return TypeXMLValue
	default:
		return TypeNone
	}
}

// NewMapping creates a new Mapping for the given session.
func NewMapping(session dtrules.Session) *Mapping {
	return &Mapping{
		session:             session,
		state:               session.GetState(),
		entities:            make(map[string]string),
		requests:            make(map[string]*EntityInfo),
		setattributes:       make(map[string]*AttributeInfo),
		multiple:            make(map[string]string),
		entitystack:         make([]string, 0),
		attribute2listPairs: make([][2]string, 0),
	}
}

// LoadMapping loads a mapping definition from an XML reader.
func (m *Mapping) LoadMapping(r io.Reader) error {
	loader := newMapLoader(m)
	return loader.Load(r)
}

// LoadData loads data from an XML reader according to the mapping.
func (m *Mapping) LoadData(r io.Reader) error {
	loader := newDataLoader(m)
	return loader.Load(r)
}

// LoadDataAndPush loads data and pushes singleton entities onto the entity stack.
// Use this for projects where singleton entities get their data from the input XML
// rather than being initialized with defaults. Entities are pushed in the order
// specified by pushOrder (e.g., []string{"state_config", "job", "taxpayer"}).
func (m *Mapping) LoadDataAndPush(r io.Reader, pushOrder []string) error {
	loader := newDataLoader(m)
	if err := loader.Load(r); err != nil {
		return err
	}
	// Push singleton entities onto the entity stack in the specified order
	for _, name := range pushOrder {
		if entity, ok := loader.entities[name]; ok {
			m.state.EntityPush(entity)
		}
	}
	return nil
}

// Initialize creates the initial entities and pushes them onto the entity stack.
func (m *Mapping) Initialize() error {
	for _, entityName := range m.entitystack {
		rname := dtrules.GetRName(entityName)
		if rname == nil {
			return fmt.Errorf("invalid entity name syntax in entity stack: %s", entityName)
		}
		entity, err := m.session.CreateEntity(rname)
		if err != nil {
			return fmt.Errorf("failed to create initial entity %s: %w", entityName, err)
		}
		m.state.EntityPush(entity)
	}
	return nil
}

// AddEntityInfo adds entity creation info for a tag.
func (m *Mapping) AddEntityInfo(tag string, info *EntityInfo) {
	m.requests[tag] = info
}

// AddAttributeInfo adds attribute mapping info for a tag.
func (m *Mapping) AddAttributeInfo(tag, enclosure, rAttribute, typeStr string) {
	aInfo := m.setattributes[tag]
	if aInfo == nil {
		aInfo = &AttributeInfo{
			Tag:     tag,
			Attribs: make(map[string]*Attrib),
		}
		m.setattributes[tag] = aInfo
	}

	attrib := &Attrib{
		Tag:        tag,
		Entity:     strings.ToLower(enclosure),
		RAttribute: strings.ToLower(rAttribute),
		Type:       ParseAttributeType(typeStr),
	}
	aInfo.Attribs[strings.ToLower(enclosure)] = attrib
}

// AddInitialEntity adds an entity to be created at initialization.
func (m *Mapping) AddInitialEntity(entityName string) {
	m.entitystack = append(m.entitystack, strings.ToLower(entityName))
}

// SetEntityCardinality sets the cardinality for an entity ("1", "*", or "+").
func (m *Mapping) SetEntityCardinality(entityName, number string) {
	m.entities[strings.ToLower(entityName)] = number
}

// AddAttribute2List adds an attribute to list mapping.
func (m *Mapping) AddAttribute2List(attribute, list string) {
	m.attribute2listPairs = append(m.attribute2listPairs,
		[2]string{strings.ToLower(attribute), strings.ToLower(list)})
}

// GetEntityInfo returns the EntityInfo for a tag.
func (m *Mapping) GetEntityInfo(tag string) *EntityInfo {
	return m.requests[tag]
}

// GetAttributeInfo returns the AttributeInfo for a tag.
func (m *Mapping) GetAttributeInfo(tag string) *AttributeInfo {
	return m.setattributes[tag]
}

// GetEntityCardinality returns the cardinality for an entity.
func (m *Mapping) GetEntityCardinality(entityName string) string {
	if c, ok := m.entities[strings.ToLower(entityName)]; ok {
		return c
	}
	return "*" // default
}

// Lookup finds the Attrib for a given enclosure, falling back to default.
func (a *AttributeInfo) Lookup(enclosure string) *Attrib {
	if attrib, ok := a.Attribs[strings.ToLower(enclosure)]; ok {
		return attrib
	}
	// Fall back to default (empty enclosure)
	return a.Attribs[""]
}
